torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,533 @@
+// Proof: According to Definition 0, (a - a % b) \ b = q is integer
+//
+// Theorem 6.5: If -|a| < r < |a|, then r % a = r, r / a = 0
+// Proof: This can be proved directly from Definition 0
+//
+// Theorem 6.6: a/(-b) = -a/b, a%(-b) = -a%b
+// Proof: See "Implementations of Div and Mod", this is a written in the paper
+//
+// Theorem 6.7: If compatible_sign(a, b), then (a + b) % c = (a % c + b % c) % c
+// where compatible_sign(a, b) is defined as ab>=0
+// Proof: According to Theorem 6.3, this is just to prove
+// a + b = a % c + b % c (mod c)
+// Because of Theorem 6.4, we have a = a % c (mod c), b = b % c (mod c),
+// applying Theorem 3.3, we get what we want
+//
+// Theorem 6.8: If compatible_sign(a, b), then (a * b) % c = (a % c * b % c) % c
+// where compatible_sign(a, b) is defined as ab>=0
+// Proof: Similar to above
+//
+// Theorem 6.9: If a is a multiple of b, then a % b = 0
+// Proof:  This can be proved directly from Euclid's division lemma
+//
+// Theorem 6.10: If b is a multiple of c, then we have: a*(b/c) = (a*b)/c
+// Proof: Same proof as 4.10
+//
+// Theorem 6.11: a/(b*c) = (a/b)/c
+// Proof: This is part of Theorem 6.1
+//
+// Theorem 6.12: a % (b * c) = a % b + ((a / b) % c) * b
+// Proof: Already proved in the proof of Theorem 6.1
+//
+// Theorem 6.13: If d divides a and b, then a % b = ((a / d) % (b / d)) * d
+// Proof: Same proof as 4.13
+//
+// Theorem 6.14: If b is a multiple of c, then a/(b/c) = (a*c)/b
+// Proof: Same proof as 4.14
+//
+// Theorem 6.15: If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+// (a+b)/c = a/c + b/c
+// Proof: From Theorem 6.7 and Theorem 6.5
+// (a + b) % c = (a % c + b % c) % c = a % c + b % c
+// The rest of the proof is the same as 4.15
+//
+// 7. Summary
+// ==========
+//
+// In summary, in nvFuser, we can use the following rules to simplify integer
+// expressions:
+//
+// A) Associativity of +: a + (b + c) = (a + b) + c
+// B) Associativity of *: a * (b * c) = (a * b) * c
+// C) Commutativity of +: a + b = b + a
+// D) Commutativity of *: a * b = b * a
+// E) Distributivity of * over +: a * (b + c) = (a * b) + (a * c)
+// F) Distributivity of * over +: (a + b) * c = (a * c) + (b * c)
+// G) (-a) / b = -(a / b) = a / (-b)
+// H) (-a) % b = -(a % b) = a % (-b)
+// I) If -|a| < r < |a|, then r % a = r, r / a = 0
+// J) Distributivity of % over +:
+//    If compatible_sign(a, b), then (a + b) % c = (a % c + b % c) % c
+// K) Distributivity of % over *:
+//    If compatible_sign(a, b), then (a * b) % c = (a % c * b % c) % c
+// L) If a is a multiple of b, then a % b = 0
+// M) If b is a multiple of c, then we have: a*(b/c) = (a*b)/c
+// N) a / (b * c) = (a / b) / c
+// O) If d divides a and b, then a % b = ((a / d) % (b / d)) * d
+// P) If b is a multiple of c, then a/(b/c) = (a*c)/b
+// Q) If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+//    (a+b)/c = a/c + b/c
+//
+// Some examples on applying the above rules to simplify expressions:
```

Comment:
Do you think canonicalization might actually be more important than simplification? I guess if possible factorization is really important, i.e. being able to try and "lift" `threadIdx.x` out of expressions as much as possible, potentially even it if means having a more complex expression.

I guess it'd be really interesting to have a simplification pass, then a canonicalization pass, CSE, then a loop hoisting pass.

Reply:
Would we for some reason even want this to be done in `IrBuilder`?

Reply:
Yes, I agree. The purpose of simplifying is actually to enable more hoisting opportunities. I just prefer to use a single word "simplification" to refer to all "simplification", "canonicalization" and "CSE". Maybe I should change my word?

Actually, I think that, in this PR, I will only implement rules A)-D) and rewrite index hoisting, and leave the rest rules to the next PR. This will make the size of this PR much smaller. With rules A)-D), I will be able to do simplifications for
```
FOR i1
  FOR i2
    FOR TIDx
      FOR BIDx
        FOR i3
          FOR i4
            index = (((i1 * 3 + i4) + 5) + ((i3 * 5 + TIDx) + 3)) + ((BIDx + i2 * 4) + 7)
```
and obtain
```
index = (((((15 + TIDx) + BIDx) + i1 * 3) + i2 * 4) + i3 * 5) + i4
```

Reply:
I would prefer to make `IrBuilder` simple and move all math-related things to a centralized place (which is the expr simplifier)